apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetService.java

```diff
@@ -248,9 +253,43 @@ public void delete(@NonNull UUID id) {
     }
 
     @Override
-    public DatasetPage find(int page, int size, DatasetCriteria criteria) {
+    public DatasetPage find(int page, int size, @NonNull DatasetCriteria criteria) {
         String workspaceId = requestContext.get().getWorkspaceId();
+        String userName = requestContext.get().getUserName();
+        String workspaceName = requestContext.get().getWorkspaceName();
+
+        if (criteria.withExperimentsOnly()) {
+
+            Mono<Set<UUID>> datasetIds = experimentDAO.findAllDatasetIds()
+                    .contextWrite(ctx -> AsyncUtils.setRequestContext(ctx, userName, workspaceName, workspaceId))
+                    .map(dto -> dto.stream()
+                            .map(ExperimentDatasetId::datasetId)
+                            .collect(toSet()));
+
+            DatasetPage datasetPage = datasetIds.flatMap(ids -> {
+
+                int maxExperimentInClauseSize = batchOperationsConfig.getDatasets().getMaxExperimentInClauseSize();
+
+                if (ids.isEmpty()) {
+                    return Mono.just(DatasetPage.empty(page));
+                } else {
+                    if (ids.size() <= maxExperimentInClauseSize) {
+                        return fetchUsingMemory(page, size, criteria, ids, workspaceId);
+                    } else {
+                        return fetchUsingTempTable(page, size, criteria, ids, workspaceId);
+                    }
```

Comment:
Have you measure the memory consumption here? Have you check if we go over the limit? If so, how big the memory increase would be needed?

As mentioned above, if this is just a memory issue I believer you're optimising prematurely, specially when increasing the memory for a Kubernetes Pod is trivial and it isn't much expensive if the amount is reasonable.

Reply:
Answered above